src/backend/distributed/utils/attribute.c

```diff
@@ -0,0 +1,621 @@
+	}
+	else
+	{
+		/*Assert(attributeToTenant == NULL); */
+	}
+
+	/*DetachSegment(); */
+
+	attributeToTenantStart = clock();
+}
+
+
+/*
+ * AnnotateQuery annotates the query with tenant attributes.
+ */
+char *
+AnnotateQuery(char *queryString, char *partitionColumn, int colocationId)
+{
+	if (partitionColumn == NULL)
+	{
+		return queryString;
+	}
+	StringInfo newQuery = makeStringInfo();
+	appendStringInfo(newQuery, ATTRIBUTE_STRING_FORMAT, partitionColumn, colocationId);
+
+	appendStringInfoString(newQuery, queryString);
+
+	return newQuery->data;
+}
+
+
+void
+CitusAttributeToEnd(QueryDesc *queryDesc)
+{
+	/*
+	 * At the end of the Executor is the last moment we have to attribute the previous
+	 * attribution to a tenant, if applicable
+	 */
+	AttributeMetricsIfApplicable();
+
+	/* now call in to the previously installed hook, or the standard implementation */
+	if (prev_ExecutorEnd)
+	{
+		prev_ExecutorEnd(queryDesc);
+	}
+	else
+	{
+		standard_ExecutorEnd(queryDesc);
+	}
+}
+
+
+/*
+ * AttributeMetricsIfApplicable updates the metrics for current tenant's statistics
+ */
+static void
+AttributeMetricsIfApplicable()
+{
+	if (strcmp(attributeToTenant, "") != 0)
+	{
+		clock_t end = { 0 };
+
+		end = clock();
+		time_t queryTime = time(0);
+		double cpu_time_used = ((double) (end - attributeToTenantStart)) / CLOCKS_PER_SEC;
+
+		if (MultiTenantMonitoringLogLevel != CITUS_LOG_LEVEL_OFF)
+		{
+			ereport(NOTICE, (errmsg("attribute cpu counter (%f) to tenant: %s",
+									cpu_time_used, attributeToTenant)));
+		}
+
+		MultiTenantMonitor *monitor = GetMultiTenantMonitor();
+
+		LWLockAcquire(&monitor->lock, LW_SHARED);
+
+		monitor->periodStart = monitor->periodStart +
+							   ((queryTime - monitor->periodStart) /
+								CitusStatsTenantsPeriod) *
+							   CitusStatsTenantsPeriod;
+
+		int tenantIndex = FindTenantStats(monitor);
+
+		if (tenantIndex == -1)
+		{
+			tenantIndex = CreateTenantStats(monitor);
+		}
+		TenantStats *tenantStats = &monitor->tenants[tenantIndex];
+
+		LWLockAcquire(&tenantStats->lock, LW_EXCLUSIVE);
+
+		UpdatePeriodsIfNecessary(monitor, tenantStats);
+		tenantStats->lastQueryTime = queryTime;
+
+		ReduceScoreIfNecessary(monitor, tenantStats, queryTime);
+
+		/*
+		 * We do this after the reducing the scores so the scores in this period are not affected by the reduction.
+		 */
+		tenantStats->score += ONE_QUERY_SCORE;
+
+
+		/*
+		 * After updating the score we might need to change the rank of the tenant in the monitor
+		 */
+		while (tenantIndex != 0 &&
+			   monitor->tenants[tenantIndex - 1].score <
+			   monitor->tenants[tenantIndex].score)
+		{
+			LWLockAcquire(&monitor->tenants[tenantIndex - 1].lock, LW_EXCLUSIVE);
+
+			ReduceScoreIfNecessary(monitor, &monitor->tenants[tenantIndex - 1],
+								   queryTime);
+
+			TenantStats tempTenant = monitor->tenants[tenantIndex];
+			monitor->tenants[tenantIndex] = monitor->tenants[tenantIndex - 1];
+			monitor->tenants[tenantIndex - 1] = tempTenant;
```

Comment:
This is copying the tenant lock. Copying locks is not an operation that's supported on them. They require that people look at the same memory locations. So if some other process is waiting on `&monitor->tenants[tenantIndex - 1].lock` while we do this swap then that's going to cause many problems.